dali/pipeline/executor/executor.cc

```diff
@@ -644,7 +667,53 @@ bool Executor<WorkspacePolicy, QueuePolicy>::HasConditionals() const {
   return has_conditionals_;
 }
 
+template<typename WorkspacePolicy, typename QueuePolicy>
+void Executor<WorkspacePolicy, QueuePolicy>::InitCheckpointing() {
+  if (!checkpointing_)
+    return;
+
+  if (std::is_same_v<QueuePolicy, SeparateQueuePolicy>)
+    DALI_FAIL("Checkpointing is not supported with `separated` pipeline exection mode enabled. ")
+
+  checkpointing_epoch_size_ = -1;
+  std::string reader_name;
+  for (const auto &node : graph_->GetOpNodes()) {
+    auto meta = node.op->GetReaderMeta();
+    if (meta.epoch_size_padded <= 0)
+      continue;
+
+    int local_epoch_size = (meta.epoch_size_padded + max_batch_size_ - 1) / max_batch_size_;
+    if (checkpointing_epoch_size_ == -1) {
+      checkpointing_epoch_size_ = local_epoch_size;
+      reader_name = node.spec.name();
+    } else if (checkpointing_epoch_size_ != local_epoch_size) {
+      DALI_FAIL(make_string(
+        "When the checkpointing is enabled, all readers must have the same epoch size. ",
+        "The readers ", reader_name, " and ", node.spec.name(), " have different epoch sizes ",
+        "(", checkpointing_epoch_size_, " and ", local_epoch_size, " respectively). "));
+    }
+  }
+
+  /* If there is no operator with ReaderMeta, set the epoch size to 1. */
+  if (checkpointing_epoch_size_ == -1)
+    checkpointing_epoch_size_ = 1;
+}
+
+template<typename WorkspacePolicy, typename QueuePolicy>
+Checkpoint &Executor<WorkspacePolicy, QueuePolicy>::GetCurrentCheckpoint() {
+  auto &cpt = GetCurrentIterationData(output_iteration_id_).checkpoint;
+  // Sanity check
+  DALI_ENFORCE(cpt.GetIterationId() == output_iteration_id_,
+               "Requested checkpoint does not correspond to the matching iteration. ");
```

Comment:
I think it may be a bit vague from the user perspective. The mismatch here simply means we did not checkpoint the given iteration and are looking at some old/default checkpoint, right? In that case, how about something like:

```suggestion
               "The pipeline cannot be checkpointed at the given iteration. Currently, checkpointing is supported at the end of the epoch only (after the last batch from the epoch was consumed).");
```